lib/IDE/TypeReconstruction.cpp

```diff
@@ -866,14 +866,27 @@ static void VisitNodeAssociatedTypeRef(
       ArchetypeType *archetype = type->getAs<ArchetypeType>();
       if (archetype) {
         Identifier identifier = ast->getIdentifier(ident->getText());
-        if (archetype->hasNestedType(identifier)) {
-          Type nested = archetype->getNestedTypeValue(identifier);
-          if (nested) {
-            result._types.push_back(nested);
-            result._module = type_result._module;
-            return;
+        Type nested;
+        if (archetype->hasNestedType(identifier))
+          nested = archetype->getNestedTypeValue(identifier);
+        else if (ProtocolDecl *self_protocol = archetype->getSelfProtocol()) {
```

Comment:
The `archetype->hasNestedType(identifier)` check should always succeed. Why are we hitting the second case?

And yes, `getSelfProtocol()` is deprecated. Please do not introduce new usages.